### ISSUE TITLE
Reindexing collection does not drop existing indexes

### ISSUE THREAD

**Issue by caseydawsonjordan** [eXist-db/exist#984]
I ran into this when debugging an index that I have, where I found that my `StreamListener` was only  receiving nodes in `ReindexMode.STORE` mode while I was expecting that a reindex would first call `NativeBroker.dropIndex()` (Triggering and index op with `ReindexMode.REMOVE_ALL_NODES`) then scan nodes for storage.

While just doing `STORE` is more efficient, it seems it would cause bugs. For instance, consider the situation where the logic for pluggable index was updated that narrowed the scope of what the index was storing. Without a `dropIndex` call first on reindex it is conceivable that data would be left in the index that is no longer valid. 

Consider the following document:

``` xml
<root>
  <one some-att="test"/>
  <two some-att="test-2"/>
  <three some-other-att="three"/>
</root>
```

For instance consider an index that operated on all nodes that contained the attribute `some-att`, the index might look like:

```
"test" => doc-id+node-id
"test-2" => doc-id+node-id
```

Then consider that index is updated to instead operate on attribute `some-other-att` and ignore `some-att`:

```
"test" => doc-id+node-id
"test-2" => doc-id+node-id
"three" => doc-id+node-id
```

Since there is no `dropIndex()` call, this index now contains two entires that should have been removed.

Is this intentional? Should I be handing this case in my index differently somehow?

Perhaps there should be a REINDEX mode so that we could still do a single pass on the document, but the index worker would understand that it could treat this as a REMOVE/STORE and handle it more appropriatly. 

In addition this would let indexes perform optimizations where they could detect data already in the storage and skip having to store it again and invalidating any caches, I can think of a few scenarios where this would result in potentially huge performance gains.

Thanks

**Comment by adamretter** [comment]
@casey Are you saying that if I store an XML document into a Collection:

```
<a>
  <b>hello</hello>
  <b>world</hello>
</a>
```

and then I was to replace it with:
```
<a>
  <b>world</hello>
</a>```

and run `reindexCollection` then the index would still contain `<b>world</b>`?

**Comment by caseydawsonjordan** [comment]
@adamretter that case would be fine because you are doing a store on the document, and stores call dropIndex() before re-indexing the new content.

The scenario that presents itself is not when the document content changes, but when the indexing logic changes. For instance, take for example if you changed your collection.xconf configuration for a given indexed and made it more restrictive, under the current paradigm old values would never be cleared out during a re-index. Only when the document content was re-stored, or the document was moved (move calls dropIndex) would the index be cleared correctly.

Another way of looking at it: The logic for dropIndex() being called in during a move operation is correct since a file could be moved into a collection with a different collection.xconf. Similarly, we could think of someone updating their collection.xconf and reindexing as needing the same logic, however that logic is missing.

**Comment by adamretter** [comment]
@caseydawsonjordan Yes, I have taken a look and I think you have found a good bug here. @wolfgangmm @ljo do you agree?

**Comment by ljo** [comment]
@adamretter Yes, I agree.
